drivers/mx25v/mx25v.c

```diff
@@ -0,0 +1,541 @@
+#define MX25V_WREN          (uint8_t)(0x06) /**< Write Enable */
+#define MX25V_WRDI          (uint8_t)(0x04) /**< Write Disable */
+#define MX25V_PP            (uint8_t)(0x02) /**< page program */
+#define MX25V_SE            (uint8_t)(0x20) /**< Sector Erase */
+#define MX25V_BE            (uint8_t)(0xD8) /**< Block Erase */
+#define MX25V_CE            (uint8_t)(0x60) /**< Chip Erase) hex code: 60 or C7 */
+#define MX25V_DP            (uint8_t)(0xB9) /**< Deep Power Down */
+#define MX25V_RDP           (uint8_t)(0xAB) /**< Release form Deep Power Down */
+/** @} */
+
+/** Timeout value for waiting operation done */
+#define MX25V_WRITE_STATUS_TIME (uint32_t)(5000)        /* write status time: TYP. 5  ms */
+#define MX25V_SECTOR_ERASE_TIME (uint32_t)(60000)       /* sector erase time: TYP. 60 ms */
+#define MX25V_BLOCK_ERASE_TIME  (uint32_t)(1000000)     /* block erase time:  TYP. 1  s */
+#define MX25V_CHIP_ERASE_TIME   (uint32_t)(15000000)    /* chip erase time:   TYP. 7  s */
+#define MX25V_PAGE_PROG_TIME    (uint32_t)(1400)        /* page program time: TYP. 1.4ms */
+
+/**
+ * @brief Send command to set or reset WEL bit.
+ *
+ * The WREN instruction is for setting Write Enable Latch (WEL) bit.
+ * The WRDI instruction is to reset Write Enable Latch (WEL) bit.
+ *
+ * @param[in] mx25v_dev     Pointer to mx25v device
+ * @param[in] cmd           WREN or WRDI command
+ *
+ * @return                  0 on operation succeeded
+ * @return                  -1 on operation failed
+ */
+static int mx25v_cmd_write_latch(mx25v_t *mx25v_dev, char cmd)
+{
+    int status;
+
+    status = spi_transfer_byte(mx25v_dev->spi_dev, cmd, NULL);
+    if (status != 1) {
+        return -1;
+    }
+
+    return 0;
+}
+
+/**
+ * @brief Check if mx25v is busy.
+ *
+ * @param[in] mx25v_dev     Pointer to mx25v device
+ *
+ * @return                  1 flash is busy (WIP bit = 1)
+ * @return                  0 flash is ready (WIP bit = 0)
+ */
+static uint8_t mx25v_is_flash_busy(mx25v_t *mx25v_dev)
+{
+    char mx25v_status;
+
+    mx25v_read_status(mx25v_dev, &mx25v_status);
+    if (mx25v_status & MX25V_WIP) {
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
+/**
+ * @brief Wait mx25v ready with timeout
+ *
+ * @param[in] mx25v_dev     Pointer to mx25v device
+ * @param[in] time_to_wait  typical time to wait for the operation
+ *
+ * @return                  1 flash is ready
+ * @return                  0 flash is time-out
+ */
+static uint8_t mx25v_wait_ready(mx25v_t *mx25v_dev, uint32_t time_to_wait)
+{
+    uint32_t start_time = xtimer_now();
+    /* the timeout is reached at 2 times of the typical operation time */
+    uint32_t timeout = start_time + (2 * time_to_wait);
+    uint32_t time = start_time;
+
+    /* sleep during typical time operation */
+    xtimer_usleep(time_to_wait);
+
+    /* wait for device ready or timeout reached */
+    while (mx25v_is_flash_busy(mx25v_dev) == 1 && time < timeout) {
+        time = xtimer_now();
```

Comment:
This busy waiting is wasting resources, it might make sense to either do another `xtimer_usleep` here.
I've seen other slower flash memories become so saturated with the SPI queries for the status bits that the chip erase took forever to complete. There's not really any reason to ask for the status again some singles microseconds after just receiving a "Write In Progress" status, better to back off and let the chip have time to finish before asking again.